### PR TITLE
Audio PR7 - Audio augmentations 2 (MFBB and Normalize)

### PR DESCRIPTION
**Pull request by SundarRajan28**
Added node and python support for ToDecibels and Normalize

### CODE REVIEW

File: rocAL/include/api/rocal_api_augmentation.h [hunk 1]
```diff
@@ -1152,4 +1152,38 @@ extern "C" RocalTensor ROCAL_API_CALL rocalSpectrogram(RocalContext p_context,
...
+ * \param [in] output_datatype the data type of the output tensor
+ * \return RocalTensor
+ */
+extern "C" RocalTensor ROCAL_API_CALL rocalToDecibels(RocalContext p_context,
+                                                      RocalTensor p_input,
+                                                      bool is_output,
+                                                      float cutoff_db,
+                                                      float multiplier,
+                                                      float reference_magnitude,
+                                                      RocalTensorOutputType rocal_tensor_output_type);
```
Comment (fiona-gladwin) [inline rocAL/include/api/rocal_api_augmentation.h]: Change rocal_tensor_output_type to output_datatype, similar to other functions

File: rocAL/include/api/rocal_api_augmentation.h [hunk 2]
```diff
@@ -1152,4 +1152,38 @@ extern "C" RocalTensor ROCAL_API_CALL rocalSpectrogram(RocalContext p_context,
                                                        int window_step,
                                                        RocalTensorOutputType output_datatype);
 
+/*! \brief A
+ * \ingroup group_rocal_augmentations
+ * \param [in] p_context Rocal context
+ * \param [in] p_input Input Rocal tensor
+ * \param [in] is_output is the output tensor part of the graph output
```
Comment (fiona-gladwin) [inline rocAL/include/api/rocal_api_augmentation.h]: Description is missing for function specific args

File: rocAL/include/api/rocal_api_augmentation.h [hunk 3]
```diff
@@ -1152,4 +1152,38 @@ extern "C" RocalTensor ROCAL_API_CALL rocalSpectrogram(RocalContext p_context,
                                                        int window_step,
                                                        RocalTensorOutputType output_datatype);
 
+/*! \brief A
```
Comment (fiona-gladwin) [inline rocAL/include/api/rocal_api_augmentation.h]: Is it incomplete here?

File: rocAL/include/api/rocal_api_augmentation.h [hunk 4]
```diff
@@ -1152,4 +1152,38 @@ extern "C" RocalTensor ROCAL_API_CALL rocalSpectrogram(RocalContext p_context,
...
+
+/*! \brief A
+ * \ingroup group_rocal_augmentations
+ * \param [in] p_context Rocal context
+ * \param [in] p_input Input Rocal tensor
+ * \param [in] is_output is the output tensor part of the graph output
+ * \param [in] output_datatype the data type of the output tensor
+ * \return RocalTensor
+ */
+extern "C" RocalTensor ROCAL_API_CALL rocalNormalize(RocalContext p_context,
```
Comment (fiona-gladwin) [inline rocAL/include/api/rocal_api_augmentation.h]: Make changes similar to previous function

File: rocAL/source/augmentations/audio_augmentations/node_normalize.cpp [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+void NormalizeNode::init(float mean, float std_dev, std::vector<int> axes, bool batch,
+                         float scale, float shift, int ddof, float epsilon) {
+    _mean = mean;
+    _std_dev = std_dev;
+    _batch = batch;
+    _scale = scale;
+    _shift = shift;
+    _ddof = ddof;
+    _epsilon = epsilon;
+    _num_of_dims = _inputs[0]->info().num_of_dims() - 1;
```
Comment (fiona-gladwin) [inline rocAL/source/augmentations/audio_augmentations/node_normalize.cpp]: This variable can be local to create_node right? Need not be a private variable

File: rocAL/source/augmentations/audio_augmentations/node_normalize.cpp [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        THROW("Adding the normalize (vxRppNormalize) node failed: " + TOSTR(status))
+}
+
+void NormalizeNode::update_node() {}
+
+void NormalizeNode::init(float mean, float std_dev, std::vector<int> axes, bool batch,
+                         float scale, float shift, int ddof, float epsilon) {
+    _mean = mean;
+    _std_dev = std_dev;
+    _batch = batch;
```
Comment (fiona-gladwin) [inline rocAL/source/augmentations/audio_augmentations/node_normalize.cpp]: Ununsed variable

File: rocAL/source/augmentations/audio_augmentations/node_normalize.cpp [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    vx_scalar shift = vxCreateScalar(vxGetContext((vx_reference)_graph->get()), VX_TYPE_FLOAT32, &_shift);
+    vx_scalar epsilon = vxCreateScalar(vxGetContext((vx_reference)_graph->get()), VX_TYPE_FLOAT32, &_epsilon);
+    vx_scalar ddof = vxCreateScalar(vxGetContext((vx_reference)_graph->get()), VX_TYPE_INT32, &_ddof);
+    vx_scalar axis_mask = vxCreateScalar(vxGetContext((vx_reference)_graph->get()), VX_TYPE_INT32, &_axis_mask);
+    _node = vxExtRppNormalize(_graph->get(), _inputs[0]->handle(), _inputs[0]->get_roi_tensor(), _outputs[0]->handle(), _outputs[0]->get_roi_tensor(), axis_mask, mean, std_dev,
+                              scale, shift, epsilon, ddof, _num_of_dims);
+
+    vx_status status;
+    if ((status = vxGetStatus((vx_reference)_node)) != VX_SUCCESS)
+        THROW("Adding the normalize (vxRppNormalize) node failed: " + TOSTR(status))
```
Comment (fiona-gladwin) [inline rocAL/source/augmentations/audio_augmentations/node_normalize.cpp]: change to vxExtRpp, same for to decibels

File: rocAL/source/api/rocal_api_augmentation.cpp [hunk 8]
```diff
@@ -2284,4 +2284,67 @@ rocalSpectrogram(
...
+        return output;
+    }
+    auto context = static_cast<Context*>(p_context);
+    auto input = static_cast<Tensor*>(p_input);
+    try {
+        RocalTensorDataType op_tensor_data_type = (RocalTensorDataType)output_datatype;
+        TensorInfo output_info = input->info();
+        output_info.set_data_type(op_tensor_data_type);
+        output = context->master_graph->create_tensor(output_info, is_output);
+        output->reset_tensor_roi();
```
Comment (fiona-gladwin) [inline rocAL/source/api/rocal_api_augmentation.cpp]: Will this output tensor and input tensor ROI differ? If it will differ or if it will be updated we can reset it

File: rocAL_pybind/amd/rocal/fn.py [hunk 9]
```diff
@@ -1109,3 +1109,33 @@ def spectrogram(*inputs, bytes_per_sample_hint = [0], center_windows = True, lay
                      "layout": layout, "power": power, "nfft": nfft, "window_length": window_length, "window_step": window_step, "rocal_tensor_output_type": rocal_tensor_output_type}
     spectrogram_output = b.Spectrogram(Pipeline._current_pipeline._handle, *(kwargs_pybind.values()))
     return spectrogram_output
+
+def to_decibels(*inputs, bytes_per_sample_hint = [0], cutoff_db = -200.0, multiplier = 10.0, reference = 0.0, seed = -1, rocal_tensor_output_type = types.FLOAT):
+    '''
+    Converts a magnitude (real, positive) to the decibel scale.
+
+    Conversion is done according to the following formula:
+
```
Comment (fiona-gladwin) [inline rocAL_pybind/amd/rocal/fn.py]: Why empty lines are insterted inbetween, please remove

File: rocAL_pybind/amd/rocal/fn.py [hunk 10]
```diff
@@ -1109,3 +1109,33 @@ def spectrogram(*inputs, bytes_per_sample_hint = [0], center_windows = True, lay
                      "layout": layout, "power": power, "nfft": nfft, "window_length": window_length, "window_step": window_step, "rocal_tensor_output_type": rocal_tensor_output_type}
     spectrogram_output = b.Spectrogram(Pipeline._current_pipeline._handle, *(kwargs_pybind.values()))
     return spectrogram_output
+
+def to_decibels(*inputs, bytes_per_sample_hint = [0], cutoff_db = -200.0, multiplier = 10.0, reference = 0.0, seed = -1, rocal_tensor_output_type = types.FLOAT):
+    '''
+    Converts a magnitude (real, positive) to the decibel scale.
```
Comment (fiona-gladwin) [inline rocAL_pybind/amd/rocal/fn.py]: converts a magnitude? doesn't sound right to me please check

File: rocAL_pybind/amd/rocal/fn.py [hunk 11]
```diff
@@ -1109,3 +1109,33 @@ def spectrogram(*inputs, bytes_per_sample_hint = [0], center_windows = True, lay
...
+
+    Conversion is done according to the following formula:
+
+    min_ratio = pow(10, cutoff_db / multiplier)
+    out[i] = multiplier * log10( max(min_ratio, input[i] / reference) )
+    '''
+    kwargs_pybind = {"input_audio0": inputs[0], "is_output": False, "cutoff_DB": cutoff_db, "multiplier": multiplier, "reference_magnitude": reference, "rocal_tensor_output_type": rocal_tensor_output_type}
```
Comment (fiona-gladwin) [inline rocAL_pybind/amd/rocal/fn.py]: What is the reason for naming input_audio0?
Can it not be input_audio, same for normalize

Change in other PR's also

Comment (fiona-gladwin) [reply rocAL_pybind/amd/rocal/fn.py]: Change rocal_tensor_output_type to output_dtype similar to image functions

File: rocAL_pybind/amd/rocal/fn.py [hunk 12]
```diff
@@ -1109,3 +1109,33 @@ def spectrogram(*inputs, bytes_per_sample_hint = [0], center_windows = True, lay
...
+    decibel_scale = b.ToDecibels(Pipeline._current_pipeline._handle, *(kwargs_pybind.values()))
+    return decibel_scale
+
+def normalize(*inputs, axes = [], axis_names = "", batch = False, bytes_per_sample_hint = [0], ddof = 0, epsilon = 0.0, mean = 0.0, scale = 1.0, seed = -1, shift = 0.0, stddev = 0.0, rocal_tensor_output_type=types.FLOAT):
+    '''
+    Normalizes the input by removing the mean and dividing by the standard deviation.
+
+    The mean and standard deviation can be calculated internally for the specified subset of axes or can be externally provided as the mean and stddev arguments.
+
```
Comment (fiona-gladwin) [inline rocAL_pybind/amd/rocal/fn.py]: Remove empty lines inbetween

File: rocAL_pybind/amd/rocal/fn.py [hunk 13]
```diff
@@ -1109,3 +1109,33 @@ def spectrogram(*inputs, bytes_per_sample_hint = [0], center_windows = True, lay
...
+    Conversion is done according to the following formula:
+
+    min_ratio = pow(10, cutoff_db / multiplier)
+    out[i] = multiplier * log10( max(min_ratio, input[i] / reference) )
+    '''
+    kwargs_pybind = {"input_audio0": inputs[0], "is_output": False, "cutoff_DB": cutoff_db, "multiplier": multiplier, "reference_magnitude": reference, "rocal_tensor_output_type": rocal_tensor_output_type}
+    decibel_scale = b.ToDecibels(Pipeline._current_pipeline._handle, *(kwargs_pybind.values()))
+    return decibel_scale
+
+def normalize(*inputs, axes = [], axis_names = "", batch = False, bytes_per_sample_hint = [0], ddof = 0, epsilon = 0.0, mean = 0.0, scale = 1.0, seed = -1, shift = 0.0, stddev = 0.0, rocal_tensor_output_type=types.FLOAT):
```
Comment (fiona-gladwin) [inline rocAL_pybind/amd/rocal/fn.py]: space before and after operator for last argument?

File: rocAL_pybind/examples/rocAL_api_audio_unittest.py [hunk 14]
```diff
@@ -66,7 +66,15 @@ def main():
...
             window_step=160,
             rocal_tensor_output_type = types.FLOAT)
-        audio_pipeline.set_outputs(spec)
+        to_decibels = fn.to_decibels(
+            spec,
+            multiplier=np.log(10),
+            reference=1.0,
+            cutoff_db=np.log(1e-20),
+            rocal_tensor_output_type=types.FLOAT,
+        )
```
Comment (fiona-gladwin) [inline rocAL_pybind/examples/rocAL_api_audio_unittest.py]: Move this bracket to previous line

File: rocAL_pybind/rocal_pybind.cpp [hunk 15]
```diff
@@ -733,5 +733,9 @@ PYBIND11_MODULE(rocal_pybind, m) {
             py::return_value_policy::reference);
     m.def("Spectrogram", &rocalSpectrogram,
             py::return_value_policy::reference);
+    m.def("audioNormalize", &rocalNormalize,"Normalizes the input by removing the mean and dividing by the standard deviation",
```
Comment (fiona-gladwin) [inline rocAL_pybind/rocal_pybind.cpp]: Space after comma

**Review by fiona-gladwin — COMMENTED** [review]
@SundarRajan28 Please address review comments

C++ unit test not present for the Audio kernels introduced in this PR

File: rocAL/CMakeLists.txt [hunk 16]
```diff
@@ -233,6 +233,7 @@ if(${BUILD_ROCAL})
                 include/augmentations/effects_augmentations/
                 include/augmentations/geometry_augmentations/
                 include/augmentations/audio_augmentations/
+                include/augmentations/arithmetic_augmentations/
```
Comment (sampath1117) [inline rocAL/CMakeLists.txt]: why arithmetic augmentations is added in this PR?

File: rocAL/include/api/rocal_api_augmentation.h [hunk 17]
```diff
@@ -1163,4 +1163,167 @@ extern "C" RocalTensor ROCAL_API_CALL rocalToDecibels(RocalContext p_context,
...
+ * \param [in] is_output is the output tensor part of the graph output
+ * \param [in] sample_hint sample_hint value is the value required to allocate the max memory for output tensor wrt resample_rate and the samples
+ * \return RocalTensor
+ */
+extern "C" RocalTensor ROCAL_API_CALL rocalResample(RocalContext p_context,
+                                                    RocalTensor p_input,
+                                                    RocalTensor p_output_resample_rate,
+                                                    RocalTensorOutputType rocal_tensor_output_type,
+                                                    bool is_output,
+                                                    float sample_hint,
```
Comment (sampath1117) [inline rocAL/include/api/rocal_api_augmentation.h]: Sample hint is float?

File: rocAL/include/api/rocal_api_augmentation.h [hunk 18]
```diff
@@ -1163,4 +1163,167 @@ extern "C" RocalTensor ROCAL_API_CALL rocalToDecibels(RocalContext p_context,
...
+
+/*! \brief Accepts F32 audio buffers
+ * \param [in] p_context Rocal context
+ * \param [in] p_input Input Rocal tensor
+ * \param [in] is_output is the output tensor part of the graph output
+ * \param [in] output_datatype the data type of the output tensor
+ * \param [in] scalar The scalar value to be multiplied with the input tensor
+ * \return RocalTensor
+ */
+extern "C" RocalTensor ROCAL_API_CALL rocalTensorMulScalar(RocalContext p_context,
```
Comment (sampath1117) [inline rocAL/include/api/rocal_api_augmentation.h]: why it is showing tensorMulScalar and tensorAddTensor as changes in this PR?

File: rocAL/include/api/rocal_api_types.h [hunk 19]
```diff
@@ -403,4 +403,24 @@ enum RocalSpectrogramLayout {
...
+    ROCAL_PAD = 0,
+    /*! \brief Trimtoshape
+     */
+    ROCAL_TRIMTOSHAPE,
+    /*! \brief Error
+     */
+    ROCAL_ERROR
+};
+
+enum RocalMelScaleFormula {
```
Comment (sampath1117) [inline rocAL/include/api/rocal_api_types.h:427]: Add description for RocalMelScaleFormula

File: rocAL/include/augmentations/arithmetic_augmentations/node_tensor_add_tensor.h [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+THE SOFTWARE.
+*/
+
+#pragma once
+#include "graph.h"
+#include "node.h"
+#include "rocal_api_types.h"
+
+class TensorAddTensorNode : public Node {
+   public:
```
Comment (sampath1117) [inline rocAL/include/augmentations/arithmetic_augmentations/node_tensor_add_tensor.h]: Please change the target branch of this PR if needed
it showing TensorAddTensor as file changes in this PR

File: rocAL/source/api/rocal_api_augmentation.cpp [hunk 21]
```diff
@@ -2279,3 +2279,305 @@ rocalToDecibels(
...
+    RocalTensorOutputType output_datatype) {
+    Tensor* output = nullptr;
+    if ((p_context == nullptr) || (p_input == nullptr)) {
+        ERR("Invalid ROCAL context or invalid input tensor")
+        return output;
+    }
+    auto context = static_cast<Context*>(p_context);
+    auto input = static_cast<Tensor*>(p_input);
+    try {
+        if ((mean > 0.0f) && (std_dev > 0.0f) && (axes.size()))
```
Comment (sampath1117) [inline rocAL/source/api/rocal_api_augmentation.cpp]: check with below condition?
` if ((mean) && (std_dev) && (axes.size()))`

File: rocAL/source/api/rocal_api_augmentation.cpp [hunk 22]
```diff
@@ -2279,3 +2279,305 @@ rocalToDecibels(
...
+            THROW("Axes must not be passed when both mean and standard deviation are specified")
+        TensorInfo output_info = input->info();
+        RocalTensorDataType op_tensor_data_type = (RocalTensorDataType)output_datatype;
+        output_info.set_data_type(op_tensor_data_type);
+        output = context->master_graph->create_tensor(output_info, is_output);
+        output->reset_tensor_roi();
```
Comment (sampath1117) [inline rocAL/source/api/rocal_api_augmentation.cpp]: Why reset_tensor_roi() is needed?

File: rocAL/source/api/rocal_api_augmentation.cpp [hunk 23]
```diff
@@ -2279,3 +2279,305 @@ rocalToDecibels(
...
+    auto context = static_cast<Context*>(p_context);
+    auto input = static_cast<Tensor*>(p_input);
+    try {
+        RocalTensorDataType op_tensor_data_type = (RocalTensorDataType)output_datatype;
+        TensorInfo output_info = input->info();
+        std::vector<size_t> max_dims = output_info.max_shape();
+        int max_frame = max_dims[0];
+        max_frame = std::max(0, max_frame);
+        std::vector<size_t> dims = output_info.dims();
+        dims[1] = max_frame;
```
Comment (sampath1117) [inline rocAL/source/api/rocal_api_augmentation.cpp]: Please check with @swetha097 
and ensure this dims assignment is correct and in sync with latest spectrogram changes

File: rocAL/source/augmentations/audio_augmentations/node_normalize.cpp (hunk 5 above)
Comment (sampath1117) [inline rocAL/source/augmentations/audio_augmentations/node_normalize.cpp]: I hope  _inputs[0]->info().num_of_dims() will be 
- 2 for 1D inputs
- 3 for 2D inputs

Please confirm

File: rocAL/source/augmentations/audio_augmentations/node_normalize.cpp [hunk 24]
```diff
@@ -0,0 +1,66 @@
...
+}
+
+void NormalizeNode::update_node() {}
+
+void NormalizeNode::init(float mean, float std_dev, std::vector<int> axes, bool batch,
```
Comment (sampath1117) [inline rocAL/source/augmentations/audio_augmentations/node_normalize.cpp]: Where are we handling case, when a vector of mean and std_dev is passed from user?

File: rocAL/source/augmentations/audio_augmentations/node_normalize.cpp [hunk 25]
```diff
@@ -0,0 +1,66 @@
...
+    _mean = mean;
+    _std_dev = std_dev;
+    _batch = batch;
+    _scale = scale;
+    _shift = shift;
+    _ddof = ddof;
+    _epsilon = epsilon;
+    _num_of_dims = _inputs[0]->info().num_of_dims() - 1;
+    if (((_mean > 0.0f) || (_std_dev > 0.0f)) && (!axes.size()))
+        std::iota(axes.begin(), axes.end(), 0);
```
Comment (sampath1117) [inline rocAL/source/augmentations/audio_augmentations/node_normalize.cpp]: if axes_size() is 0, then only this code will be executed
but axes.begin() and axes.end() will be invalid right if vector size is 0?

Please check this

File: rocAL/source/augmentations/audio_augmentations/node_normalize.cpp [hunk 26]
```diff
@@ -0,0 +1,66 @@
...
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+*/
+
+#include "node_normalize.h"
+
+#include <vx_ext_rpp.h>
+
+#include "exception.h"
```
Comment (sampath1117) [inline rocAL/source/augmentations/audio_augmentations/node_normalize.cpp]: Remove empty blank lines between includes

File: rocAL_pybind/amd/rocal/fn.py [hunk 27]
```diff
@@ -1110,3 +1110,88 @@ def to_decibels(*inputs, bytes_per_sample_hint = [0], cutoff_db = -200.0, multip
...
+
+    The mean and standard deviation can be calculated internally for the specified subset of axes or can be externally provided as the mean and stddev arguments.
+
+    The normalization is done following the formula:
+
+    out = scale * (in - mean) / stddev + shift
+
+    The formula assumes that out and in are equally shaped tensors, but mean and stddev might be either tensors of same shape, scalars, or a mix of these.
+    '''
+    kwargs_pybind = {"input_audio0": inputs[0], "is_output": False, "batch": batch, "axes": axes, "mean": mean, "stddev": stddev,
```
Comment (sampath1117) [inline rocAL_pybind/amd/rocal/fn.py]: it was mentioned as "input_audio0" here again
in Resample PR also it was the same case

Please check and correct all such instances to input_audio

File: rocAL_pybind/rocal_pybind.cpp [hunk 28]
```diff
@@ -141,6 +141,8 @@ std::unordered_map<int, std::string> rocalToPybindOutputDtype = {
     {1, "float16"},
     {2, "uint8"},
     {3, "int8"},
+    {4, "uint32"},
+    {5, "int32"},
```
Comment (sampath1117) [inline rocAL_pybind/rocal_pybind.cpp]: why this change is added in this PR?

**Review by sampath1117 — COMMENTED** [review]
Please address review comments

**Comment by fiona-gladwin** [comment]
@SundarRajan28 Please clean up /rebase this PR

File: rocAL/source/augmentations/audio_augmentations/node_normalize.cpp (hunk 26 above)
Comment (SundarRajan28) [reply rocAL/source/augmentations/audio_augmentations/node_normalize.cpp]: This is formatting changes done by vscode formatter

File: rocAL/include/api/rocal_api_augmentation.h [hunk 29]
```diff
@@ -1277,4 +1277,52 @@ extern "C" RocalTensor ROCAL_API_CALL rocalSlice(RocalContext context,
                                                  RocalOutOfBoundsPolicy policy = RocalOutOfBoundsPolicy::ROCAL_ERROR,
                                                  RocalTensorOutputType output_datatype = ROCAL_FP32);
 
+/*! \brief Performs mean-stddev normalization on images.
```
Comment (fiona-gladwin) [inline rocAL/include/api/rocal_api_augmentation.h:1280]: Is it only on images?

File: rocAL/include/api/rocal_api_augmentation.h [hunk 30]
```diff
@@ -1277,4 +1277,52 @@ extern "C" RocalTensor ROCAL_API_CALL rocalSlice(RocalContext context,
...
+ * \ingroup group_rocal_augmentations
+ * \param [in] context Rocal context
+ * \param [in] input Input Rocal tensor
+ * \param [in] axes axes list for tensor normalization
+ * \param [in] mean mean value (specified for each channel) for tensor normalization
+ * \param [in] std_dev standard deviation value (specified for each channel) for tensor normalization
+ * \param [in] scale scale value (specified for each channel) for tensor normalization
```
Comment (fiona-gladwin) [inline rocAL/include/api/rocal_api_augmentation.h:1287]: scale and shift is also specified in each channel?

File: rocAL/include/api/rocal_api_augmentation.h [hunk 31]
```diff
@@ -1277,4 +1277,52 @@ extern "C" RocalTensor ROCAL_API_CALL rocalSlice(RocalContext context,
...
+                                                     std::vector<unsigned> &axes,
+                                                     std::vector<float> &mean,
+                                                     std::vector<float> &std_dev,
+                                                     bool is_output,
+                                                     float scale = 1.0, float shift = 0.0,
+                                                     RocalTensorOutputType output_datatype = ROCAL_FP32);
+
+/*! \brief Applies mel-filter bank augmentation on the given input tensor
+ * \ingroup group_rocal_augmentations
+ * \param [in] p_context Rocal context
```
Comment (fiona-gladwin) [inline rocAL/include/api/rocal_api_augmentation.h:1304]: Name p_context and p_input as context and input, follow the same thing in all API

File: rocAL/include/augmentations/audio_augmentations/node_mel_filter_bank.h [hunk 32]
```diff
@@ -0,0 +1,45 @@
+/*
+Copyright (c) 2019 - 2023 Advanced Micro Devices, Inc. All rights reserved.
```
Comment (fiona-gladwin) [inline rocAL/include/augmentations/audio_augmentations/node_mel_filter_bank.h:2]: Change copyright to 2024

File: rocAL/source/augmentations/audio_augmentations/node_mel_filter_bank.cpp [hunk 33]
```diff
@@ -0,0 +1,59 @@
+/*
+Copyright (c) 2019 - 2023 Advanced Micro Devices, Inc. All rights reserved.
```
Comment (fiona-gladwin) [inline rocAL/source/augmentations/audio_augmentations/node_mel_filter_bank.cpp:2]: Change copyright to 2024

File: rocAL/source/augmentations/audio_augmentations/node_mel_filter_bank.cpp [hunk 34]
```diff
@@ -0,0 +1,59 @@
...
+
+MelFilterBankNode::MelFilterBankNode(const std::vector<Tensor *> &inputs, const std::vector<Tensor *> &outputs) : Node(inputs, outputs) {}
+
+void MelFilterBankNode::create_node() {
+    if (_node)
+        return;
+
+    vx_scalar freq_high = vxCreateScalar(vxGetContext((vx_reference)_graph->get()), VX_TYPE_FLOAT32, &_freq_high);
+    vx_scalar freq_low = vxCreateScalar(vxGetContext((vx_reference)_graph->get()), VX_TYPE_FLOAT32, &_freq_low);
+    vx_scalar mel_formula = vxCreateScalar(vxGetContext((vx_reference)_graph->get()), VX_TYPE_UINT32, &_mel_formula);
```
Comment (fiona-gladwin) [inline rocAL/source/augmentations/audio_augmentations/node_mel_filter_bank.cpp:37]: Mel formula is an enum better to keep it as an int32, instead of uint32

File: rocAL/include/augmentations/effects_augmentations/node_normalize.h [hunk 35]
```diff
@@ -0,0 +1,46 @@
...
+
+   protected:
+    void create_node() override;
+    void update_node() override{};
+
+   private:
+    int _axis_mask = 0;
+    uint _compute_mean, _compute_stddev;
+    vx_array _mean_vx_array, _stddev_vx_array;
+    std::vector<unsigned> _axes;
```
Comment (fiona-gladwin) [inline rocAL/include/augmentations/effects_augmentations/node_normalize.h:42]: All variables in lines 40-42, can be made local

File: rocAL/include/augmentations/effects_augmentations/node_normalize.h [hunk 36]
```diff
@@ -0,0 +1,46 @@
...
+    std::vector<float> _mean, _std_dev;
+    float _scale, _shift;
+    std::vector<std::vector<uint32_t>> _normalize_roi;
```
Comment (fiona-gladwin) [inline rocAL/include/augmentations/effects_augmentations/node_normalize.h:45]: Make it local variable in create_node if required

File: rocAL/source/augmentations/effects_augmentations/node_normalize.cpp [hunk 37]
```diff
@@ -0,0 +1,95 @@
...
+    status |= vxAddArrayItems(_mean_vx_array, mean_vec.size(), mean_vec.data(), sizeof(vx_float32));
+    if (status != 0)
+        THROW(" vxAddArrayItems failed in the normalize node (vxExtRppNormalize)  node: " + TOSTR(status) + "  " + TOSTR(status))
+
+    _stddev_vx_array = vxCreateArray(vxGetContext((vx_reference)_graph->get()), VX_TYPE_FLOAT32, stddev_vec.size());
+    status |= vxAddArrayItems(_stddev_vx_array, stddev_vec.size(), stddev_vec.data(), sizeof(vx_float32));
+    if (status != 0)
+        THROW(" vxAddArrayItems failed in the normalize node (vxExtRppNormalize)  node: " + TOSTR(status) + "  " + TOSTR(status))
+
+    vx_scalar axis_mask = vxCreateScalar(vxGetContext((vx_reference)_graph->get()), VX_TYPE_INT32, &_axis_mask);
```
Comment (fiona-gladwin) [inline rocAL/source/augmentations/effects_augmentations/node_normalize.cpp:75]: Add _vx suffix to all vx_scalar variables

File: rocAL/source/augmentations/effects_augmentations/node_normalize.cpp [hunk 38]
```diff
@@ -0,0 +1,95 @@
...
+    _compute_stddev = _std_dev.size() ? 0 : 1;
+
+    int mean_stddev_array_size = 1;
+    auto nDim = _inputs[0]->info().num_of_dims() - 1;
+    uint axis[nDim];
+    for (unsigned i = 0; i < _batch_size; i++) {
+        int totalElements = 1;
+        unsigned *tensor_shape = _inputs[0]->info().roi()[i].end;
+        for (uint j = 0; j < nDim; j++) {
+            axis[j] = ((_axis_mask & (int)(pow(2, j))) >= 1) ? 1 : 0;
```
Comment (fiona-gladwin) [inline rocAL/source/augmentations/effects_augmentations/node_normalize.cpp:45]: Better to add comment, as to what is done here

File: rocAL_pybind/amd/rocal/fn.py [hunk 39]
```diff
@@ -1164,3 +1164,26 @@ def slice(*inputs, anchor = [], shape = [], fill_values = [0.0],  out_of_bounds_
...
+    '''
+    kwargs_pybind = {"input_tensor": inputs[0], "axes": axes, "mean": mean, "stddev": stddev, "is_output": False,
+                     "scale": scale, "shift": shift, "output_datatype": output_datatype}
+    normalize_output = b.normalize(Pipeline._current_pipeline._handle, *(kwargs_pybind.values()))
+    return normalize_output
+
+def mel_filter_bank(*inputs, bytes_per_sample_hint = [0], freq_high = 0.0, freq_low = 0.0, mel_formula = types.SLANEY, nfilter = 128, normalize = True, sample_rate = 44100.0, seed = -1, output_datatype = types.FLOAT):
+    '''
+    Converts a spectrogram to a mel spectrogram by applying a bank of triangular filters.
+    The frequency ('f') dimension is selected from the input layout. In case of no layout, “f”, “ft”, or “*ft” is assumed, depending on the number of dimensions.
```
Comment (fiona-gladwin) [inline rocAL_pybind/amd/rocal/fn.py:1184]: 'ft' is mentioned twice, is it supposed to be 'tf'?

File: rocAL_pybind/examples/rocAL_api_audio_unittest.py [hunk 40]
```diff
@@ -300,16 +343,20 @@ def main():
...
-            audio_pipeline = non_silent_region_and_slice(batch_size=batch_size, num_threads=num_threads, device_id=device_id, rocal_cpu=rocal_cpu, path=audio_path, file_list=file_list)
-        if case_name == "non_silent_region":
-            audio_pipeline = non_silent_region(batch_size=batch_size, num_threads=num_threads, device_id=device_id, rocal_cpu=rocal_cpu, path=audio_path, file_list=file_list)
         if case_name == "resample":
             audio_pipeline = resample_pipeline(batch_size=batch_size, num_threads=num_threads, device_id=device_id, rocal_cpu=rocal_cpu, path=audio_path, file_list=file_list)
         if case_name == "tensor_add_tensor":
             audio_pipeline = tensor_add_tensor_pipeline(batch_size=batch_size, num_threads=num_threads, device_id=device_id, rocal_cpu=rocal_cpu, path=audio_path, file_list=file_list)
         if case_name == "tensor_mul_scalar":
             audio_pipeline = tensor_mul_scalar_pipeline(batch_size=batch_size, num_threads=num_threads, device_id=device_id, rocal_cpu=rocal_cpu, path=audio_path, file_list=file_list)
+        if case_name == "slice":
```
Comment (fiona-gladwin) [inline rocAL_pybind/examples/rocAL_api_audio_unittest.py:356]: Why slice is added here?

File: rocAL/source/api/rocal_api_augmentation.cpp [hunk 41]
```diff
@@ -2504,3 +2504,68 @@ rocalSlice(
...
+        return output;
+    }
+    auto context = static_cast<Context*>(p_context);
+    auto input = static_cast<Tensor*>(p_input);
+    try {
+        RocalTensorDataType op_tensor_data_type = (RocalTensorDataType)output_datatype;
+        TensorInfo output_info = input->info();
+        std::vector<size_t> max_dims = output_info.max_shape();
+        int max_frame = max_dims[0];
+        max_frame = std::max(0, max_frame);
```
Comment (fiona-gladwin) [inline rocAL/source/api/rocal_api_augmentation.cpp]:
```suggestion
        int max_frame = max_dims[0];
        max_frame = std::max(0, max_frame);
```

Do this in a single line

File: rocAL/source/api/rocal_api_augmentation.cpp [hunk 42]
```diff
@@ -2504,3 +2504,68 @@ rocalSlice(
...
+    RocalTensorOutputType output_datatype) {
+    Tensor* output = nullptr;
+    if ((p_context == nullptr) || (p_input == nullptr)) {
+        ERR("Invalid ROCAL context or invalid input tensor")
+        return output;
+    }
+    auto context = static_cast<Context*>(p_context);
+    auto input = static_cast<Tensor*>(p_input);
+    try {
+        RocalTensorDataType op_tensor_data_type = (RocalTensorDataType)output_datatype;
```
Comment (fiona-gladwin) [inline rocAL/source/api/rocal_api_augmentation.cpp:2554]: Use static cast

File: rocAL/source/api/rocal_api_augmentation.cpp [hunk 43]
```diff
@@ -2504,3 +2504,68 @@ rocalSlice(
...
+        TensorInfo output_info = input->info();
+        std::vector<size_t> max_dims = output_info.max_shape();
+        int max_frame = max_dims[0];
+        max_frame = std::max(0, max_frame);
+        std::vector<size_t> dims = output_info.dims();
+        dims[1] = max_frame;
+        dims[2] = nfilter;
+        output_info.set_dims(dims);
+        output_info.set_data_type(op_tensor_data_type);
```
Comment (fiona-gladwin) [inline rocAL/source/api/rocal_api_augmentation.cpp:2563]: This supports only float data type right?
Add condition and I don't think this will be required, since the input will laready be float type

File: rocAL/source/api/rocal_api_augmentation.cpp [hunk 44]
```diff
@@ -2504,3 +2504,68 @@ rocalSlice(
...
+    if ((p_context == nullptr) || (p_input == nullptr)) {
+        ERR("Invalid ROCAL context or invalid input tensor")
+        return output;
+    }
+    auto context = static_cast<Context*>(p_context);
+    auto input = static_cast<Tensor*>(p_input);
+    try {
+        RocalTensorDataType op_tensor_datatype = static_cast<RocalTensorDataType>(output_datatype);
+        TensorInfo output_info = input->info();
+        output_info.set_data_type(op_tensor_datatype);
```
Comment (fiona-gladwin) [inline rocAL/source/api/rocal_api_augmentation.cpp:2523]: Are all data types supported in Normalize?
